### ISSUE TITLE
libpldm/platform: Using unions for variable-length fields is incompatible with DSP0248

### ISSUE THREAD

**Issue by FHomps** [openbmc/pldm#29]
https://github.com/openbmc/pldm/blob/ef773059fdead2135c96c4a4c3520e4752012ef0/libpldm/platform.h#L426:L455

Following DMTF DSP0248-1.2.1 (and earlier), some PDR fields are of variable length depending on value types specified earlier in the PDR. For example, in section 28.11, table 87:
![image](https://user-images.githubusercontent.com/33905666/152507227-209805f0-0d48-4b26-8af3-74bb0cd95f3f.png)

This maxSettable field (and others) should be of a length determined by the effecterDataSize field present earlier in the PDR:
![image](https://user-images.githubusercontent.com/33905666/152507361-8e9aaadb-d8a8-41b5-94ce-eb07cf01366d.png)

Using unions for such a purpose is not possible, as unions always have the size of their biggest member; GCC's `__attribute__((packed))` does not change this. This effectively makes all PDRs following this architecture incompatible with the DMTF specification.

**Comment by FHomps** [comment]
I thought about trying to fix this transparently by adding PDR type checks to `encode_get_pdr_resp()` and `decode_get_pdr_resp()`, and copying the PDR structure into the PLDM message piece by piece, removing the unwanted padding when needed.

However, this will not work, due to caller ownership of the message (for `encode_get_pdr_resp()`) and of the record data (for `decode_get_pdr_resp()`) in the current API. The client needs to initialize both those structures and is expected to know their size before calling the functions.

Fixing this issue will thus require changes to the API. I see three possible options to solve this:
1. Completely rework the PDR system to use getters and setters / dynamic pointer offsets for member access, with underlying opaque structures of variable size representing the PDRs
2. Grant ownership of the problematic variable size structures to the encode and decode functions, requiring the client to free them afterwards
3. Create a `compute_pdr_transmission_size()` function (or equivalent) to be called before the encode / decode functions. The record data / message size would then depend on the returned dynamic size.

I am personally very much against option 2, and have a preference for option 3. However, all three do break the current API or current usage of the library.

**Comment by selvamv-create** [comment]
Can someone please let know what is the final resolution made for this issue ? Any spec update or openbmc changes made ?
Im working on a feature where the use case is requiring different sizes for the sensor.
Like 16/32/64 bit sensor/effecter to be supported, in that case what will be the sizes for min/max settable and other union ranges in the sensor.
Can someone please respond, Thanks !!!

**Comment by amboar** [comment]
> Using unions for such a purpose is not possible, as unions always have the size of their biggest member; GCC's __attribute__((packed)) does not change this. This effectively makes all PDRs following this architecture incompatible with the DMTF specification.

Right. `__attribute__((packed))` is not a viable solution. Rather, the solution is in the form of several recommendations in `CONTRIBUTING.md`:

1. [New public `struct` definitions are _not_ marked `__attribute__((packed))`](https://github.com/openbmc/libpldm/blob/4d006e6578bb9a3aaa250170401afc6df0c2b5d5/CONTRIBUTING.md#L159)
2. [If the work interacts with the PLDM wire format, then it does so using the `msgbuf` APIs found in `src/msgbuf.h` (and under `src/msgbuf/`) to minimise concerns around spatial memory safety and endian-correctness](https://github.com/openbmc/libpldm/blob/4d006e6578bb9a3aaa250170401afc6df0c2b5d5/CONTRIBUTING.md#L232C6-L234C77)

Essentially, library `encode_*()` and `decode_*()` APIs must be used to exchange data between message fields and message structs defined by the library. We throw away the suggestion that casting struct definitions over message buffers is workable by requiring that new structs added to the library don't specify `__attribute__((packed))`. This also improves ergonomics for library users, as structs marked `__attribute__((packed))` carry some unfortunate constraints.

Implementation of the `encode_*()` and `decode_*()` APIs must be done in terms of the msgbuf APIs (internal to the library) to ease review and maintenance.

Some work towards fixing instances of this problem has been done. For example, `struct pldm_numeric_effecter_value_pdr` [contains members of type `union_effecter_data_size`](https://github.com/openbmc/libpldm/blob/4d006e6578bb9a3aaa250170401afc6df0c2b5d5/include/libpldm/platform.h#L820). [`decode_numeric_effecter_pdr_data()` implements the extraction](https://github.com/openbmc/libpldm/blob/main/src/dsp/platform.c#L2972-L3065) in a way that caters for various member sizings specified in the received message.

**Comment by selvamv-create** [comment]
> > Using unions for such a purpose is not possible, as unions always have the size of their biggest member; GCC's **attribute**((packed)) does not change this. This effectively makes all PDRs following this architecture incompatible with the DMTF specification.
> 
> Right. `__attribute__((packed))` is not a viable solution. Rather, the solution is in the form of several recommendations in `CONTRIBUTING.md`:
> 
> 1. [New public `struct` definitions are _not_ marked `__attribute__((packed))`](https://github.com/openbmc/libpldm/blob/4d006e6578bb9a3aaa250170401afc6df0c2b5d5/CONTRIBUTING.md#L159)
> 2. [If the work interacts with the PLDM wire format, then it does so using the `msgbuf` APIs found in `src/msgbuf.h` (and under `src/msgbuf/`) to minimise concerns around spatial memory safety and endian-correctness](https://github.com/openbmc/libpldm/blob/4d006e6578bb9a3aaa250170401afc6df0c2b5d5/CONTRIBUTING.md#L232C6-L234C77)
> 
> Essentially, library `encode_*()` and `decode_*()` APIs must be used to exchange data between message fields and message structs defined by the library. We throw away the suggestion that casting struct definitions over message buffers is workable by requiring that new structs added to the library don't specify `__attribute__((packed))`. This also improves ergonomics for library users, as structs marked `__attribute__((packed))` carry some unfortunate constraints.
> 
> Implementation of the `encode_*()` and `decode_*()` APIs must be done in terms of the msgbuf APIs (internal to the library) to ease review and maintenance.
> 
> Some work towards fixing instances of this problem has been done. For example, `struct pldm_numeric_effecter_value_pdr` [contains members of type `union_effecter_data_size`](https://github.com/openbmc/libpldm/blob/4d006e6578bb9a3aaa250170401afc6df0c2b5d5/include/libpldm/platform.h#L820). [`decode_numeric_effecter_pdr_data()` implements the extraction](https://github.com/openbmc/libpldm/blob/main/src/dsp/platform.c#L2972-L3065) in a way that caters for various member sizings specified in the received message.

Thanks for the response, its helpful.
